### PR TITLE
feat(opencode): add HTTP/HTTPS proxy support

### LINKED ISSUE #10227: add proxy support
### Description

Hi team,

when I use opencode behind an enterprise proxy I'm getting this error
```
(base) PS C:\Users\s494709\AppData\Local\OpenCode> .\opencode-cli.exe
{
  "name": "UnknownError",
  "data": {
    "message": "Error: self signed certificate in certificate chain"
  }
}
```

Can you please add proxy support?

### Plugins

_No response_

### OpenCode version

latest

### Steps to reproduce

_No response_

### Screenshot and/or share link

(base) PS C:\Users\s494709\AppData\Local\OpenCode> .\opencode-cli.exe
{
  "name": "UnknownError",
  "data": {
    "message": "Error: self signed certificate in certificate chain"
  }
}

### Operating System

Windows 11

### Terminal

powershell

### PR DESCRIPTION
**Pull request by Thesam1798**
Fixes #10227
Related: #9736, #8822

## Summary

Bun's native `fetch()` ignores `HTTP_PROXY`/`HTTPS_PROXY` environment variables. This PR adds a `proxyFetch` wrapper that reads proxy configuration and passes it to Bun's fetch.

## Changes

- Add `ProxyConfig` schema in config.ts
- Add `proxyFetch` wrapper in `src/util/fetch.ts`
- Replace all `fetch()` calls with `proxyFetch()` across ~20 files
- Support `NO_PROXY` patterns (wildcards, suffixes)
- Add `OPENCODE_DISABLE_PROXY` emergency bypass flag
- Add unit and integration tests
- Add documentation in `docs/proxy.md`

## Configuration

Environment variables:
```bash
export HTTPS_PROXY=http://proxy:8080
export NO_PROXY=localhost,*.internal.com
```

Or in `opencode.json`:
```json
{
  "proxy": {
    "https": "http://proxy:8080",
    "no_proxy": ["localhost", "*.internal.com"]
  }
}
```

## Verification

- All 822 existing tests pass
- Tested manually behind corporate proxy (France Travail)
- Verified API calls go through proxy
- Verified `NO_PROXY` bypass works for internal hosts

### CODE REVIEW

File: packages/opencode/test/util/fetch.test.ts [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    delete process.env.http_proxy
+    delete process.env.https_proxy
+    delete process.env.no_proxy
+    delete process.env.OPENCODE_DISABLE_PROXY
+
+    // Clear config
+    setProxyConfig(undefined)
+  })
+
+  afterEach(() => {
```
[the remark this answers is not in the document]
Comment (Thesam1798) [reply packages/opencode/test/util/fetch.test.ts:22]: 1. **`beforeEach` already handles cleanup** - Every test starts by explicitly deleting all proxy-related env vars before each test runs
2. **Bun test isolation** - Bun runs test files in separate workers, so env var leakage between FILES is not a concern
3. **Practical impact is negligible** - Given the existing `beforeEach` pattern

File: packages/opencode/src/config/config.ts [hunk 2]
```diff
@@ -979,6 +995,9 @@ export namespace Config {
         )
         .optional()
         .describe("MCP (Model Context Protocol) server configurations"),
+      proxy: ProxyConfig.optional().describe(
+        "HTTP/HTTPS proxy configuration. Overrides HTTP_PROXY, HTTPS_PROXY, and NO_PROXY environment variables.",
+      ),
```
[the remark this answers is not in the document]
Comment (Thesam1798) [reply packages/opencode/src/config/config.ts:1134]: Critical bug fixed! Added `setProxyConfig(result.proxy)` call in the config state() function before returning.

The proxy config from opencode.json is now properly wired to the fetch layer.

File: packages/opencode/test/util/fetch-integration.test.ts [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+describe("proxy-fetch integration", () => {
+  // Save original env vars
+  const originalEnv = { ...process.env }
+
+  afterEach(() => {
+    // Restore original env vars
+    Object.assign(process.env, originalEnv)
+    // Clear config
+    setProxyConfig(undefined)
+  })
```
[the remark this answers is not in the document]
Comment (Thesam1798) [reply packages/opencode/test/util/fetch-integration.test.ts:41]: Implemented - added cleanup loop to remove env vars introduced during tests before restoring originals.

File: packages/opencode/test/util/fetch-integration.test.ts [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+      process.env.OPENCODE_DISABLE_PROXY = "1"
+
+      expect(getProxyForUrl("https://example.com")).toBeUndefined()
+    })
+
+    test("respects falsy disable values", () => {
+      process.env.HTTPS_PROXY = "http://proxy:8080"
+      process.env.OPENCODE_DISABLE_PROXY = "0"
+
+      // "0" is truthy as a string, so proxy should be disabled
```
[the remark this answers is not in the document]
Comment (Thesam1798) [reply packages/opencode/test/util/fetch-integration.test.ts]: Good catch! Renamed to "disables proxy when flag is any non-empty string" and updated comment to match actual behavior.

File: packages/opencode/docs/proxy.md [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+   ```bash
+   echo $NO_PROXY
+   ```
+
+### SSL/TLS errors through proxy
+
+Some corporate proxies perform SSL interception. You may need to:
+
+1. Configure the proxy's CA certificate in your system trust store
+2. Set `NODE_TLS_REJECT_UNAUTHORIZED=0` (not recommended for production)
```
[the remark this answers is not in the document]
Comment (Thesam1798) [reply packages/opencode/docs/proxy.md]: Valid security concern. Removed `NODE_TLS_REJECT_UNAUTHORIZED=0` and replaced with proper solutions:
- `NODE_EXTRA_CA_CERTS` for additional CA certificates
- `BUN_OPTIONS="--use-system-ca"` for system certificate store
- New `proxy.tls.ca` config option (added in next commit)

File: packages/opencode/src/util/fetch.ts [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+ *
+ * // Explicitly set proxy for this request
+ * const response = await proxyFetch("https://api.example.com", {
+ *   proxy: "http://other-proxy:8080"
+ * })
+ * ```
+ */
+export async function proxyFetch(
+  input: RequestInfo | URL,
+  init?: BunFetchInit & { proxy?: string | false | { url: string; headers?: Record<string, string> } },
```
[the remark this answers is not in the document]
Comment (Thesam1798) [reply packages/opencode/src/util/fetch.ts]: Fixed in commit 4dffe0b1d - now using `Omit<BunFetchInit, "proxy">` as suggested.

**Comment by Thesam1798** [comment]
**This PR (#10856)** implements:
- System-wide HTTP/HTTPS proxy support via env vars and `opencode.json`
- `NO_PROXY` pattern matching (wildcards, suffixes, exact match)
- `proxyFetch()` wrapper for Bun's native fetch
- TLS configuration for proxy connections (`proxy.tls.ca`, `proxy.tls.rejectUnauthorized`)

**PR #6025** focuses on:
- Per-provider TLS settings (client certs, mTLS)
- Per-provider proxy overrides

The two PRs are complementary - this one provides the global proxy infrastructure, while #6025 will add provider-specific overrides on top of it.

File: packages/opencode/src/util/fetch.ts [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+  }
+
+  if (tls.ca) {
+    const files = Array.isArray(tls.ca) ? tls.ca : [tls.ca]
+    // Validate paths don't contain traversal
+    for (const p of files) {
+      if (p.includes("..")) {
+        throw new Error(`Invalid CA path (path traversal not allowed): ${p}`)
+      }
+    }
```
[the remark this answers is not in the document]
Comment (Thesam1798) [reply packages/opencode/src/util/fetch.ts:143]: This is intentional. The path traversal check is a simple security measure for defense-in-depth. Users who need relative paths with `..` can use absolute paths instead. This limitation is acceptable since:
1. It's a user-provided config value, not untrusted input
2. Absolute paths work everywhere
3. The edge case (needing `../` in CA paths) is rare
We can add a comment documenting this limitation.

File: packages/opencode/test/util/fetch-integration.test.ts [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+// Skip integration tests in CI or when no proxy is configured
+const hasProxy = !!(
+  process.env.HTTP_PROXY ||
+  process.env.HTTPS_PROXY ||
+  process.env.http_proxy ||
+  process.env.https_proxy
+)
+const isCI = !!(process.env.CI || process.env.GITHUB_ACTIONS || process.env.GITLAB_CI)
+const shouldSkip = isCI || !hasProxy
+
```
[the remark this answers is not in the document]
Comment (Thesam1798) [reply packages/opencode/test/util/fetch-integration.test.ts:25]: The current gating is sufficient:
- Tests are skipped in CI (`GITHUB_ACTIONS`, `GITLAB_CI`, etc.)
- Tests only run when `HTTP_PROXY`/`HTTPS_PROXY` are explicitly set
Developers with proxy env vars in their shell are typically in corporate environments where proxy awareness is expected. Adding another opt-in flag creates redundant friction. The `shouldSkip = isCI || !hasProxy` logic adequately protects against unintended execution.

File: packages/opencode/src/util/fetch.ts [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+  return {
+    http: _configProxy?.http || process.env.HTTP_PROXY || process.env.http_proxy,
+    https: _configProxy?.https || process.env.HTTPS_PROXY || process.env.https_proxy,
+    noProxy:
+      _configProxy?.no_proxy ||
+      (process.env.NO_PROXY || process.env.no_proxy || "")
+        .split(",")
+        .map((s) => s.trim())
+        .filter(Boolean),
+  }
```
[the remark this answers is not in the document]
Comment (Thesam1798) [reply packages/opencode/src/util/fetch.ts:59]: Good catch, but this is low priority. The edge case (setting `http: ""` or `no_proxy: []` to explicitly disable while env vars are set) is unlikely in practice:
- Empty string is an invalid proxy URL
- `no_proxy: []` is functionally equivalent to not setting it
- `OPENCODE_DISABLE_PROXY` exists for explicit bypass
This can be addressed in a future cleanup if needed.

File: packages/opencode/docs/proxy.md [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+  "proxy": {
+    "https": "http://proxy:8080",
+    "tls": {
+      "rejectUnauthorized": false
+    }
+  }
+}
+```
+
+> **Warning:** Setting `rejectUnauthorized: false` disables TLS certificate validation for proxy connections. This makes the connection vulnerable to man-in-the-middle attacks. Only use this for trusted internal proxies where you cannot install the CA certificate.
```
[the remark this answers is not in the document]
Comment (Thesam1798) [reply packages/opencode/docs/proxy.md:185]: This is already addressed in the documentation:
1. CA-based solutions are shown **first** and labeled "(recommended)"
2. Multiple CA examples are provided before the `rejectUnauthorized` option
3. The `rejectUnauthorized: false` section includes a prominent warning about MITM vulnerability
4. It's clearly documented as a "use with caution" last resort
The documentation structure already follows best practices by steering users toward proper CA configuration first.

**Comment by Thesam1798** [comment]
## Real-world feedback

I encountered this issue with a corporate proxy. Without this fix, opencode was completely unusable:
- **All providers** failed (couldn't reach APIs)
- **At startup**: version checks failed, ripgrep download failed, etc.

After this fix, opencode works **perfectly** without any special configuration - the `HTTP_PROXY`/`HTTPS_PROXY` environment variables are automatically detected.

While waiting for reviews on this PR, I'm using a locally built version and everything works as expected!

**Comment by andyWang1688** [comment]
Thanks for this PR! I’m a Desktop user behind a corporate proxy and can’t reliably inject env vars into GUI apps. A config-file proxy (opencode.json) would solve this for me. This is blocking adoption on my side—any chance to merge soon?

**Comment by Thesam1798** [comment]
I'll check this week, but I think the latest versions of Bun and Opencode fix this!

If so, I'll close this PR, otherwise I'll continue to update it from my Dev branch.

**Comment by Thesam1798** [comment]
I tested the native **1.2.6** version, and the issue has been resolved.
